[cmd3] Add a declarative state machine API on top of commands v3

SamCarlberg opened the pull request:
This provides an API for writing a finite state machine compatible with the commands v3 framework. Individual states in the state machine are wrappers around command objects (which may themselves be state machines). Transitions between states are defined with a staged builder DSL similar to command builders, and uses `@NoDiscard` to catch partially configured transitions.

The FSM API is meant to handle highly complex cases that the fluent command chaining DSL and coroutine-based imperative commands cannot easily represent; specifically, where a command sequence may want to go back to an arbitrary previous state or skip forward to an arbitrary future state.

Here's an example from the design doc for a command that will drive to a known scoring location, aim at a scoring target, and repeatedly shoot balls until a storage hopper is empty. It also has conditions to stop shooting and move back to the scoring location if it's jostled away, and then automatically resume firing.

```java
public Command autoWithStateMachine() {
  // Declare the state machine
  StateMachine stateMachine = new StateMachine("Auto With State Machine");

  // Define states
  State getInPosition = stateMachine.addState(drivetrain.driveToScoringLocation());
  State aiming = stateMachine.addState(turret.aimAtGoal());
  State scoring = stateMachine.addState(shooter.fireOnce());
  State celebrating = stateMachine.addState(leds.celebrate());

  // Set the initial state. Neglecting this will cause a runtime exception when the state machine starts.
  // Teams using the WPILib compiler plugin will get a compiler error if they do not set this
  stateMachine.setInitialState(getInPosition);

  // Switch to aiming when we reach the scoring location.
  getInPosition.switchTo(aiming).whenComplete();
  // Set the swerve wheels in an X shape after reaching the scoring location to resist being pushed away.
  getInPosition.onExit(() -> Scheduler.getDefault().fork(drivetrain.setX()));

  // Then start scoring once the turret is aimed at the goal.
  aiming.switchTo(scoring).when(turret::aimedAtGoal);

  // Loop the scoring state as long as the hopper has a ball.
  scoring.switchTo(scoring).whenCompleteAnd(() -> hopper.hasBall());

  // Automatically interrupt any part of the aiming or scoring sequence if
  // the robot is moved away from the scoring location and move back into position.
  stateMachine.switchFromAny(aiming, scoring).to(getInPosition).when(atScoringLocation.negate());

  // Start celebrating once the final ball has been scored.
  scoring.switchTo(celebrating).whenCompleteAnd(() -> !hopper.hasBall());

  return stateMachine;
}
```

A compiler check is added to detect object construction that's not followed by post-construction initializer methods (as defined by the class by placing `@PostConstructionInitializer` on such methods). `StateMachine.setInitialState` uses this to detect team code that creates a state machine but does not set its initial state.

SamCarlberg committed:
Add a declarative state machine API on top of commands

Daniel1464 commented on commandsv3/src/main/java/org/wpilib/commands3/StateMachine.java:
```diff
@@ -0,0 +1,475 @@
+   * @param name The name of the state machine. Cannot be null. This will appear in telemetry as the
+   *     {@link Command#name() name} of the state machine.
+   */
+  public StateMachine(String name) {
```
It's better to pass in the initial state command in the constructor(to minimize runtime errors) and to fetch it with fsm.getInitialState() or something of that sort

SamCarlberg commented:
> When the javac plugin gets merged, a plugin should be added that makes sure the return types of switchTo(), switchFromAny(), etc. aren't dropped.

This is already in place. The `@NoDiscard` annotations on the builder classes are enforcing this

SamCarlberg committed:
Fix wrong method name in javadocs

SamCarlberg committed:
Add `@PostConstructionInitializer` annotation

Used to annotate methods that are required to be called after constructing a new object

Methods with this annotation must be called directly on the field or variable; the plugin can't check for calls on objects returned by methods, or on objects returned by ternary expressions or similar

SamCarlberg committed:
Add `@PostConstructionInitializer` to `StateMachine.setInitialState`

Makes usages that forget to set the initial state a compiler error, rather than just a runtime error

Runtime error checking is left in place in case teams build their code without the compiler plugin enabled

SamCarlberg committed:
Linting

SamCarlberg committed:
Update PostConstructionInitializer to allow static methods

Add an annotation processor to validate definitions of static initializer methods (requiring exactly one unambiguous input argument)

DRY out suppression detections

SamCarlberg committed:
Linting

SamCarlberg committed:
Fix compiler plugin tests after rebase

SamCarlberg committed:
Merge branch '2027' into commandsv3-state-machines

SamCarlberg committed:
Merge remote-tracking branch 'wpi/2027' into commandsv3-state-machines

SamCarlberg committed:
Merge branch '2027' into commandsv3-state-machines

Daniel1464 commented:
> > Change whenComplete() to whenDone()(conveys the same meaning in less characters).
> 
> I'm not sure if I like this. It reads a little weirdly, and is a little awkward with `whenDoneAnd`

This might be a bit late, but could you explain why this reads weirdly? `whenDoneAnd(condition)` seems fine to me.

SamCarlberg committed:
Merge branch '2027' into commandsv3-state-machines

SamCarlberg committed:
Make `runningCommands` return a list instead of a set

Ensures ordering, useful for tracing parent -> child relationships

SamCarlberg committed:
Improve end condition transitions

Calling `.whenComplete()` multiple times will have the last call win, by virtue of ejecting prior `.whenComplete()` bindings. Otherwise, `.whenCompleteAnd()` transitions will always take precedence over general `.whenComplete()` ones. Completion transitions are evaluated in declaration order

SamCarlberg committed:
Merge branch '2027' into commandsv3-state-machines

SamCarlberg committed:
Fix typo in docs.

EdanThomton commented:
Not sure how possible this would be, but 
```java
state1.switchTo(state2).whenCompleteAnd(cond).otherwiseSwitchTo(state3)
```
method would be really nice for if/else branching logic. It doesn't need to look exactly like this, something like this would also work
```java
state1.whenComplete().switchTo(state2).if(cond).otherwiseSwitchTo(state3)
// or this if you want less verbose
state1.whenComplete().switchIfCondOtherwise(cond, state2, state3)
```
The latter two would need a rework of `.whenComplete()`, as `.whenComplete()` would need to be a factory method on a given state rather than a switch object

Another idea is this, which would read more like traditional logic
```java
state1.whenComplete()
    .if(cond).switchTo(state2)
    .elseIf(cond2).switchTo(state3)
    .else().switchTo(state4)
```

SamCarlberg commented:
> if/else branching logic

A supplier would be better. User code could do whatever it wants, instead of being limited to the DSL we build.

```java
state1.switchTo(() -> {
  if (cond) {
    return state2;
  } else if (cond2) {
    return state3;
  } else {
    return state4;
  }
}).whenComplete();
```


The conditions could be moved to a helper method to make the state machine definition cleaner

```java
state1.switchTo(() -> nextAfterState1()).whenComplete();

State nextAfterState1() {
  if (cond) {
    return state2;
  } else if (cond2) {
    return state3;
  } else {
    return state4;
  }
}
```

EdanThomton commented:
Actually a supplier would be awesome. That also allows for good switch statements and other complex logic.

SamCarlberg committed:
Add support for dynamic state switching

Instead of hardcoding everything ahead of time in a DSL, allow users to optionally provide suppliers that will be evaluated at transition time to dynamically determine the next state to move to. Useful if there are multiple branches off a single state so users don't need to compose multivariate boolean expressions

SamCarlberg committed:
Lint

SamCarlberg committed:
Add javadoc and fix incorrect error string

JosephTLockwood commented:
This is probably outside the scope of this initial pull request, but something that I could see being useful is a way to name transitions. This would leave the door open for printing or generating graphs of the StateMachine, allowing users to visualize the full state machine structure. For example, visualizing your original example shows how things can really start to grow with just a few states:

<img width="661" height="637" alt="image" src="https://github.com/user-attachments/assets/b14e1039-658d-44c8-b4d1-afec6f8784fb" />

This also raises a concern about the branching logic example. While your approach is easier to read, if named transitions and graph introspection become goals, the dynamic lambda pattern makes that impossible since the graph structure is unknowable at construction time

> 
> ```java
> state1.switchTo(() -> nextAfterState1()).whenComplete();
> 
> State nextAfterState1() {
>   if (cond) {
>     return state2;
>   } else if (cond2) {
>     return state3;
>   } else {
>     return state4;
>   }
> }
> ```

Would have to be changed to something like this.

```java
private State nextState;
State scoring = stateMachine.addState(coroutine -> {
    if (cond)       nextState = state2;
    else if (cond2) nextState = state3;
    else            nextState = state4;
});
stateMachine.switchOnResult(scoring, () -> nextState);
```

With a possible addition to StateMachine:

```java
public void switchOnResult(State from, Supplier<State> result) {
    for (State target : registeredStates) {
        // .named(target.getName()) could auto-name transitions if State exposes a name
        from.switchTo(target).whenCompleteAnd(() -> result.get() == target);
    }
}
```

This is admittedly less readable than the dynamic lambda approach, but it provides three concrete benefits:

1. The graph is fully declared upfront, so the StateMachine object knows all possible destinations at construction time, enabling visualization and logging.
2. Transitions are explicit, so if a named transitions API is added, they automatically get labels.
3. It prevents accidentally returning a state that isn't registered in the machine, which would be a silent runtime bug in the dynamic lambda approach.
4. (A side effect benefit) - You aren't manually creating a state machine inside of your state machine.

Arnaud-5G commented:
> ```java
> state1.switchTo(() -> {
>   if (cond) {
>     return state2;
>   } else if (cond2) {
>     return state3;
>   } else {
>     return state4;
>   }
> }).whenComplete();
> ```

I feel like that removes the utility of the `whenCompleteAnd()` function by replacing it with a more verbose and less intuitive approach.

What was there before: (not sure if this would work exactly like I think it does but it should be the exact same behavior as the one below)
```java
state1.switchTo(state2).whenCompleteAnd(()-> cond1);
state1.switchTo(state3).whenCompleteAnd(()-> cond2);
state1.switchTo(state4).whenComplete();
```

What it will look like with `switchTo(Supplier)`:
```java
state1.switchTo(()-> {
   if(cond1) {
      return state2;
   }
   if(cond2) {
      return state3;
   } else if (!cond1 && !cond2) {
      return state4;
   }
}).whenComplete();
```

I feel like the first solution would be best for code readability and beginner friendliness. The second option feels like a `Commands.defer()` which in my experience is hard for beginners to grasp and understand how to do lambda blocks correctly. Simple lambdas are way easier to understand.

SamCarlberg commented:
> I feel like that removes the utility of the whenCompleteAnd() function by replacing it with a more verbose and less intuitive approach.

`whenCompleteAnd` still exists, and will still be the recommended way of doing most conditional transitions. But in cases where there are potentially many conditions, checking each permutation exactly and exhaustively can be unwieldy and error prone.

EdanThomton commented:
To second what Sam said, whenCompleteAnd will not go away. The switchTo supplier is only for complex branching logic that requires it (such as if/else branches when complete, see my original post on it)

EdanThomton commented:
On the topic of named transitions, it would be a cool thing to have, but I don't see it happening in java without majorly reworking the API, especially with the switchTo supplier.

JosephTLockwood commented:
Correct, it would not work with a `switchTo` supplier. You would have to structure it the way I showed. Which, as I mentioned, has the con of being less readable to newer programmers. I just wanted to bring it up, as I feel named translations could be very useful for debugging and potential visualizations by doing a somewhat minor code restructuring.

Daniel1464 commented:
Why not have when/whenComplete be chainable? I.E

```
state.switchTo(state2).when(conditionA).when(conditionB);
```

Daniel1464 commented:
> This is probably outside the scope of this initial pull request, but something that I could see being useful is a way to name transitions. This would leave the door open for printing or generating graphs of the StateMachine

Good news - i just got a gradle plugin working that generates mermaid graphs of methods returning finite state machines (and annotated with the @GenerateDiagram annotation). It's still a heavy WIP, but it seems to function pretty well!

https://github.com/frc-5160-the-chargers/Teaching2027/tree/main/buildSrc/src/main/java/frc/gradle
